File: csharp/MetaObjects.Conformance.Tests/Navigator.cs
// Navigator.cs — port of typescript/packages/metadata/test/conformance/navigator.ts.
//
// Interprets a script.json navigate path over the typed MetaData tree.
// A path segment is either `type:name` or `type[subType]` for nameless nodes.

using System.Text.RegularExpressions;
using MetaObjects.Meta;

namespace MetaObjects.Conformance.Tests;

/// <summary>
/// Walks a <see cref="MetaData"/> tree along a sequence of path segments.
/// </summary>
public static class Navigator
{
    private static readonly Regex BracketPattern =
        new(@"^([a-z]+)\[([a-zA-Z]+)\]$", RegexOptions.Compiled);

    /// <summary>
    /// Walk <paramref name="root"/>.Children() matching each path segment.
    /// Returns the resolved node, or <see langword="null"/> if any segment misses.
    /// An empty path returns <paramref name="root"/> itself (identity case).
    /// </summary>
    public static MetaData? Navigate(MetaData root, IReadOnlyList<string> path)
    {
        MetaData current = root;
        foreach (var segment in path)
        {
            var next = current.Children().FirstOrDefault(c => MatchSegment(c, segment));
            if (next is null) return null;
            current = next;
        }
        return current;
    }

    private static bool MatchSegment(MetaData node, string segment)
    {
        var bracketMatch = BracketPattern.Match(segment);
        if (bracketMatch.Success)
        {
            return node.Type == bracketMatch.Groups[1].Value &&
                   node.SubType == bracketMatch.Groups[2].Value;
        }

        var colonIdx = segment.IndexOf(':');
        if (colonIdx == -1) return false;

        return node.Type == segment[..colonIdx] &&
               node.Name == segment[(colonIdx + 1)..];
    }
}

File: typescript/packages/metadata/test/conformance/navigator.ts
// navigator.ts — interprets a script.json navigate path over the typed
// MetaData tree. A path segment is `type:name`, or `type[subType]` for a
// nameless node.

import type { MetaData } from "../../src/meta/meta-data.js";

function matchSegment(node: MetaData, segment: string): boolean {
  const bracket = segment.match(/^([a-z]+)\[([a-zA-Z]+)\]$/);
  if (bracket) return node.type === bracket[1] && node.subType === bracket[2];
  const colon = segment.indexOf(":");
  if (colon === -1) return false;
  return node.type === segment.slice(0, colon) && node.name === segment.slice(colon + 1);
}

/** Walk `root.children()` matching each path segment; undefined if any segment misses. */
export function navigate(root: MetaData, path: readonly string[]): MetaData | undefined {
  let current: MetaData = root;
  for (const segment of path) {
    const next = current.children().find((c) => matchSegment(c, segment));
    if (next === undefined) return undefined;
    current = next;
  }
  return current;
}
